python/drned_xmnr/op/filtering.py

```diff
@@ -356,184 +355,312 @@ def event_generator(consumer):
+
     def handle(self, event):
-        if isinstance(event, DrnedCommitEvent):
-            event.mark_complete()
-            return None
-        if not isinstance(event, DrnedActionEvent):
-            return TERMINATED
-        event.mark_complete()
-        if event.action == 'commit':
-            return (event.produce_line(), DrnedCommitState())
-        elif event.action == 'compare_config':
-            return (event.produce_line(), DrnedCompareState())
-        else:
-            return event.produce_line()
+        if isinstance(event, StartState) or isinstance(event, Transition):
+            return (True,
+                    [GenS(DrnedPrepare), TransitionS(),
+                     GenS(TransFailed), GenS(InitFailed),
+                     self],
+                    event.produce_line())
+        if isinstance(event, DrnedTeardown):
+            return (True, [ActionS('restore')], event.produce_line())
+        return (False, [])
 
 
-class DrnedCommitState(LogState):
+class GenS(LS):
+    '''A generic log state.
+
+    Every event of given type is accepted (and its line is produced),
+    everything else is left for further processing.
+
+    (So e.g. GenS(prepare) -> prepare | [])
+
+    '''
+
+    def __init__(self, eventclass):
+        self.eventclass = eventclass
+        self.name = 'gen({})'.format(eventclass.__name__)
+
     def handle(self, event):
-        if isinstance(event, DrnedCommitLogEvent):
-            event.mark_complete()
-            return DrnedCommitResultState()
-        elif isinstance(event, DrnedCommitNNEvent):
-            event.mark_complete()
-            return DrnedCommitNNState()
-        elif isinstance(event, DrnedEmptyCommit):
-            event.mark_complete()
-            return (event.produce_line(), TERMINATED)
-        elif isinstance(event, DrnedCommitEvent):
-            event.mark_complete()
-            return TERMINATED
-        return TERMINATED
-
-
-class DrnedCommitResultState(LogState):
+        if isinstance(event, self.eventclass):
+            return (True, [], event.produce_line())
+        return (False, [])
+
+
+class WalkTransitionsS(LS):
+    '''
+    WT -> (pytest Transition)* (teardown Restore Commit Compare)?
+    '''
+
+    name = 'transitions'
+
     def handle(self, event):
-        if isinstance(event, DrnedCommitResult):
-            event.mark_complete()
-            if event.success:
-                return (event.produce_line(), TERMINATED)
-            else:
-                return DrnedFailureState()
-        return TERMINATED
+        if isinstance(event, PyTest):
+            return (True, [TransitionS(), self], event.produce_line())
+        if isinstance(event, DrnedTeardown):
+            return (True,
+                    [GenS(DrnedRestore),
+                     ActionS('commit', [CommitS()]),
+                     ActionS('compare_config', [CompareS()])], event.produce_line())
+
+        return (False, [])
+
 
+class SilentTransitionsS(LS):
```

Comment:
Isn't this name a bit confusing? It's SilentTransitionS bus the listed states are the one for which output is produced...

Reply:
I guess it is; perhaps "OverviewTransitions", or do you have better idea? The point is that the class is used when the log-detail is "overview" and the filtered output should be considerably shorter - in particular, a transition A->B is mentioned in only one line (if there are no failures), no details such as commits, comparisons, rollbacks.